kroxylicious/src/main/java/io/kroxylicious/proxy/internal/ApiVersionsServiceImpl.java

```diff
@@ -3,32 +3,42 @@
  *
  * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
  */
-package io.kroxylicious.proxy.internal.filter;
+
+package io.kroxylicious.proxy.internal;
 
 import java.util.HashSet;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 
+import org.apache.kafka.common.message.ApiVersionsRequestData;
 import org.apache.kafka.common.message.ApiVersionsResponseData;
-import org.apache.kafka.common.message.ResponseHeaderData;
+import org.apache.kafka.common.message.ApiVersionsResponseData.ApiVersion;
+import org.apache.kafka.common.message.ApiVersionsResponseData.ApiVersionCollection;
 import org.apache.kafka.common.protocol.ApiKeys;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import io.kroxylicious.proxy.filter.ApiVersionsResponseFilter;
+import io.kroxylicious.proxy.ApiVersionsService;
 import io.kroxylicious.proxy.filter.FilterContext;
-import io.kroxylicious.proxy.filter.ResponseFilterResult;
 
-/**
- * Changes an API_VERSIONS response so that a client sees the intersection of supported version ranges for each
- * API key. This is an intrinsic part of correctly acting as a proxy.
- */
-public class ApiVersionsFilter implements ApiVersionsResponseFilter {
+public class ApiVersionsServiceImpl {
+
+    private record ApiVersions(ApiVersionCollection upstream, ApiVersionCollection intersected) {
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ApiVersionsServiceImpl.class);
+    private ApiVersions apiVersions = null;
 
-    private static final Logger LOGGER = LoggerFactory.getLogger(ApiVersionsFilter.class);
+    public void mutateVersionsApplyingIntersection(String channel, ApiVersionsResponseData upstreamApiVersions) {
+        var upstream = upstreamApiVersions.duplicate().apiKeys();
+        intersectApiVersions(channel, upstreamApiVersions);
+        var intersected = upstreamApiVersions.duplicate().apiKeys();
+        this.apiVersions = new ApiVersions(upstream, intersected);
+    }
```

Comment:
The method name seems a bit funky and I think the side effect is a bit strange, but I guess its a requirement to expose this as a public method...

Maybe it should just be `updateVersions`? Mutation is implied and intersection is an implementation detail?

Reply:
Yeah, update is probably a good enough convention to show we are mutating